core/services/workflows/engine.go

```diff
@@ -158,6 +130,90 @@ LOOP:
 	e.logger.Info("engine initialized")
 }
 
+func (e *Engine) initializeCapability(ctx context.Context, s *step, retrySec int) error {
+	// If the capability already exists, that means we've already registered it
+	if s.capability != nil {
+		return nil
+	}
+
+	cp, innerErr := e.registry.Get(ctx, s.Type)
+	if innerErr != nil {
+		return fmt.Errorf("failed to get capability with ref %s: %s, retrying in %d seconds", s.Type, innerErr, retrySec)
+	}
+
+	// We only need to configure actions, consensus and targets here, and
+	// they all satisfy the `CallbackCapability` interface
+	cc, ok := cp.(capabilities.CallbackCapability)
+	if !ok {
+		return fmt.Errorf("could not coerce capability %s to CallbackCapability", s.Type)
+	}
+
+	if s.config == nil {
+		configMap, ierr := values.NewMap(s.Config)
+		if ierr != nil {
+			return fmt.Errorf("failed to convert config to values.Map: %s", ierr)
+		}
+		s.config = configMap
+	}
+
+	reg := capabilities.RegisterToWorkflowRequest{
+		Metadata: capabilities.RegistrationMetadata{
+			WorkflowID: e.workflow.id,
+		},
+		Config: s.config,
+	}
+
+	innerErr = cc.RegisterToWorkflow(ctx, reg)
+	if innerErr != nil {
+		return fmt.Errorf("failed to register to workflow (%+v): %w", reg, innerErr)
+	}
+
+	s.capability = cc
+	return nil
+}
+
+// initializeExecutionStrategy for `step`.
+// Broadly speaking, we'll use `immediateExecution` for non-target steps
+// and `scheduledExecution` for targets. If we don't have the necessary
+// config to initialize a scheduledExecution for a target, we'll fallback to
+// using `immediateExecution`.
+func (e *Engine) initializeExecutionStrategy(step *step) error {
+	if step.executionStrategy != nil {
+		return nil
+	}
+
+	ie := immediateExecution{}
+	if step.Type != capabilities.CapabilityTypeTarget.String() {
+		step.executionStrategy = ie
+		return nil
+	}
+
+	if e.donInfo == nil {
+		step.executionStrategy = ie
+		return nil
+	}
+
+	var position *int
+	for i, w := range e.donInfo.Peers {
+		if w == e.donInfo.PeerID {
+			idx := i
+			position = &idx
+		}
+	}
+
+	if position == nil {
+		step.executionStrategy = ie
+		return nil
+	}
+
+	step.executionStrategy = scheduledExecution{
+		sharedSecret: e.donInfo.SharedSecret,
+		n:            len(e.donInfo.Peers),
```

Comment:
Continuing on the other thread here:
This is incorrect for remote targets - If workflow DON had N nodes but a capability has M, we should be iterating over M, not N. Still each workflow node will do the calculation but then send it to the selected remote one to execute.

Reply:
I'm going to work through remote targets in more detail, so to make some forward progress will include only the local changes for review to start with.